src/reporter.cc

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+#include "reporter.h"
+
+#define BOOST_NETWORK_ENABLE_HTTPS
+#include <boost/network/protocol/http/client.hpp>
+
+#include "configuration.h"
+#include "format.h"
+#include "json.h"
+#include "logging.h"
+#include "time.h"
+
+namespace http = boost::network::http;
+
+namespace google {
+
+MetadataReporter::MetadataReporter(const MetadataAgentConfiguration& config,
+                                   MetadataStore* store, double period_s)
+    : store_(store),
+      config_(config),
+      environment_(config_),
+      auth_(environment_),
+      period_(period_s),
+      reporter_thread_([=]() { ReportMetadata(); }) {}
+
+MetadataReporter::~MetadataReporter() {
+  reporter_thread_.join();
+}
+
+void MetadataReporter::ReportMetadata() {
+  LOG(INFO) << "Metadata reporter started";
+  // Wait for the first collection to complete.
+  // TODO: Come up with a more robust synchronization mechanism.
+  std::this_thread::sleep_for(time::seconds(3));
+  // TODO: Do we need to be able to stop this?
+  while (true) {
+    if (config_.VerboseLogging()) {
+      LOG(INFO) << "Sending metadata request to server";
+    }
+    try {
+      SendMetadata(store_->GetMetadataMap());
+      if (config_.VerboseLogging()) {
+        LOG(INFO) << "Metadata request sent successfully";
+      }
+      if (config_.MetadataReporterPurgeDeleted()) {
+        store_->PurgeDeletedEntries();
+      }
+    } catch (const boost::system::system_error& e) {
+      LOG(ERROR) << "Metadata request unsuccessful: " << e.what();
+    }
+    std::this_thread::sleep_for(period_);
+  }
+  //LOG(INFO) << "Metadata reporter exiting";
```

Comment:
I'd vote for uncommenting this, so I can create a log metric, but please remove if you disagree.

Reply:
I'd like to keep this PR focused on the refactoring. Happy to do it in a follow-up PR — there are a few other places where extra logging would help.

Reply:
FWIW, I think this specific bit follows an infinite loop, so will never be printed. It's commented out pending the TODO above the loop.